Environment

etoledom opened the pull request:
This PR adds a new global called `Environment`. 

The `Environment` is meant to gather all globals / singletons used in the app that define the global state of the "world". In this way, is easier to take control of our app Environment when we need to Test.

Ideally we can start implementing and using this `Environment` in small steps:

- If we need to access a global value or send a message to a global entity, we should try to reach it through the `Environment.current`. 
- If it doesn't exist we can add it to the Environment:
  - First we declare a protocol that define the global's interface (its enough to declare just what you need, or the most common use cases)
  - Then we make the global conform to that interface. 
  - Finally we can add a property to the Environment following the example of the properties already there.

**For testing:**

We can create doubles that implement the needed protocol, and then we call `Environment.replaceEnvironment(valueToReplace: MyDouble())`.

**To test this PR:**
- Build and run.
- Make sure that the changes on `NotificationDetailsViewController.swift` where the `Environment` is used are working properly (it calls the object it should).

frosty commented on WordPress/Classes/Utility/Environment/Environment.swift:
```diff
@@ -0,0 +1,47 @@
+}
+
+extension Environment {
+    /// Creates a new Environment, changin just a subset of the current global dependencies.
```
Nitpick, typo: `changin` -> `changing`

frosty commented on WordPress/Classes/Utility/Environment/Environment.swift:
```diff
@@ -0,0 +1,47 @@
+    let appRatingUtility: AppRatingUtilityType
+
+    /// The main Core Data context
+    let mainContext: NSManagedObjectContext
```
I wonder if the Environment should provide the `ContextManager`, instead of the context itself 🤔 

The call site would then become: `Environment.current.contextManager.mainContext`.

etoledom replied:
I was thinking the same thing. `ContextManager` has methods that we will need to be able to test anyway. The `Environment.current.contextManager.mainContext` (so long) was what made me rethink that.

Do you think it is just fine? 

As an alternative we could add a helper computed var to access `mainContext` directly, so we can keep `Environment.current.mainContext`.

jleandroperez replied:
Caveat here is: the Services architecture requires you to... well, pass over a MOC during initialization. If you wanted to go BG, you'd find yourself recreating some of the goodies we've got in ContextManager. 

Alternatively: You can always do...

```
let contextManager: ContextManager

var mainContext: NSManagedObjectContext {
   return contextManager.mainContext
}
```

etoledom commented:
Thanks @frosty and @jleandroperez !

I added `contextManagerType` to the list to hold a type conforming to the `ContextManagerType` protocol. It defaults to `ContextManager.self`.

Following @jleandroperez comment, I added a computed property `mainContext` to save us from `Environment.current.contextManagerType.shared.mainContext`.
It becomes `Environment.current.mainContext` as before.

The current declaration of `ContextManagerType` has just the `mainContext` property, but it can easily be extended to add the many creation and save context methods.

I followed the convention (from KickStarter) of adding `-Type` to the property name when it holds a type instead of an instance (they have a couple of those too).

frosty commented on WordPress/Classes/Utility/Environment/Environment.swift:
```diff
@@ -0,0 +1,53 @@
+
+    private init(
+        appRatingUtility: AppRatingUtilityType = AppRatingUtility.shared,
+        contextManagerType: ContextManagerType.Type = ContextManager.self) {
```
I'm a little confused – why don't we handle context manager the same way as `appRatingUtility`? So `shared` isn't part of the protocol, and we just have `ContextManager.shared` as the default value here.

etoledom replied:
You are completely right! I got confused. 😅
This is useful for when the class has static methods and this is not the case. I'll fix it right now.

etoledom commented:
@frosty - ready for another look!

jleandroperez commented:
Thanks for adding the calculated property @etoledom !!! looks great!!
